packages/webrtc-core/src/media/track.ts

```diff
@@ -0,0 +1,198 @@
+/* eslint-disable no-underscore-dangle */
+import { EventMap } from 'typed-emitter';
+import { EventEmitter } from '../event-emitter';
+import { logger } from '../util/logger';
+import { MediaStreamTrackKind } from '../peer-connection';
+
+export enum Events {
+  Ended = 'ended',
+  Muted = 'muted',
+  /**
+   * Fires when the published state of a LocalTrack changes.  A value of 'true' indicates the
+   * track has been published, and 'false' indicates that it is no longer published.  Tracks are
+   * unpublished by default, and must be published explicitly (no event will be fired to indicate
+   * the initial state of 'unpublished').
+   */
+  PublishedStateUpdate = 'published-state-update',
+  /**
+   * Fires when there has been a change in the underlying track.
+   */
+  UnderlyingTrackChange = 'underlying-track-change',
+}
+
+export interface TrackState {
+  id: string;
+  label: string;
+  muted: boolean;
+}
+
+export type TrackEndEvent = {
+  trackState: TrackState;
+};
+
+export type TrackMuteEvent = {
+  trackState: TrackState;
+};
+// eslint-disable-next-line no-shadow
+export enum TrackStatus {
+  ENDED = 'ended',
+  LIVE = 'live',
+}
+
+// eslint-disable-next-line no-shadow
+export enum TrackKind {
+  AUDIO = 'audio',
+  VIDEO = 'video',
+}
+
+export type TrackPublishEvent = {
+  isPublished: boolean;
+  trackState: TrackState;
+};
+
+export interface TrackEvents extends EventMap {
+  [Events.Ended]: (event: TrackEndEvent) => void;
+  [Events.Muted]: (event: TrackMuteEvent) => void;
+  [Events.UnderlyingTrackChange]: () => void;
+}
+
+export type TrackEffect = unknown;
+// TBD: Fix this once types are published separately
+// export type TrackEffect = BaseMicrophoneEffect | BaseCameraEffect;
+
+/**
+ * Basic Track class.
+ */
+export abstract class Track extends EventEmitter<TrackEvents> {
+  static Events = Events;
+
+  ID: string;
+
+  kind: MediaStreamTrackKind;
+
+  status: TrackStatus;
+
+  label: string;
+
+  #mediaStreamTrack: MediaStreamTrack;
+
+  #mediaStreamTrackWithEffect?: MediaStreamTrack;
+
+  /**
+   * Constructor for the Track class. Creates an empty CoreLocalTrack or uses an existing one.
+   *
+   * @param track - Media stream track.
+   */
+  constructor(track: MediaStreamTrack) {
+    super();
+    this.ID = track.id;
+    this.status = track.readyState as TrackStatus;
+    this.label = track.label;
+    this.kind = track.kind as MediaStreamTrackKind;
+    this.#mediaStreamTrack = track;
+
+    /**
+     * Emit ended event when the underlying track ends.
+     */
+    this.#mediaStreamTrack.onended = () => {
+      this.emit(Events.Ended, { trackState: this.trackState });
+    };
+
+    /**
+     * Emit mute event when the underlying track gets muted.
+     */
+    this.#mediaStreamTrack.onmute = () => {
+      this.emit(Events.Muted, {
+        trackState: {
```

Comment:
in `this.#mediaStreamTrack.onended` we just use `this.trackState`, should we do the same here?